lib/api/src/exports.rs

```diff
@@ -150,6 +150,19 @@ impl Exports {
             .map_err(|_| ExportError::IncompatibleType)
     }
 
+    /// Hack to get this working with nativefunc too
+    pub fn get_with_generics<'a, T, Args, Rets>(&'a self, name: &str) -> Result<T, ExportError>
```

Comment:
`get_native_function_with_generics` ?

Reply:
This lets us use the `derive(WasmerEnv)` with `NativeFunc`

The issue is that `instance.exports.get()` does not work for `NativeFunc` because `NativeFunc` has generic arguments that need to be known.  This is an alternative `get` function that _does_ work with `NativeFunc` because it takes generics as arguments.

This way we have a single entrypoint that works for all the types we support: we're offloading the work of deciding what logic to run to the typesystem / traits.  Without this we can't handle `NativeFunc` without special casing it in the macro (that is, we have to have users explicitly call it out for us because we need to decide which type of code to generate. This is what I had before with the `#[wasmer(export(native_func("hello")))]` syntax.)

Reply:
I meant that we should rename it to `get_native_function_with_generics` :P

Reply:
Oh, we can but we're calling it for all types not just `NativeFunc`, it's `Exportable` / `Instance::get` + NativeFunc support, it works for `Memory`, `Table`, and everything `Exportable` too, which is why it's useful!

Reply:
In the `get` function we are using `Exportable<'a>`. However in this function decl we are using `Args` and `Rets` which kind of implies a function? (or that's why I assumed it was for a function).

Why would we want this function working for `Memory` and `Table`?

Reply:
The reason is that we now have a single function by which can talk about anything that can come from an `Instance`, whether it has `Args` and `Rets` or not.

It was created to solve the problem of the procedural macro not knowing anything about what was tagged with `#[export]`.  Before that if we wanted to get `NativeFunc` with it's `Args` and `Rets` we had to call it out explicitly in the macro: `#[export(native_func)]`.  The new API lets us not care what it is because we have a single interface by which we can get anything.

Reply:
Ok, got it.

Reply:
Should we leave this API public only for the crate? (but not for outside)
This might prevent unintended usage for now

Reply:
Well, it needs to be public to be called by the macro-expanded code: this is the primary interface that the macro will use when it expands